metricq/interval_source.py

```diff
@@ -143,33 +143,27 @@ async def task(self) -> None:
                 # task from being cancelled.
                 await self.stop(e)
 
-            try:
-                if self._period is None:
-                    raise ValueError(
-                        "IntervalSource.period not set before running task"
-                    )
+            if self._period is None:
+                raise ValueError("IntervalSource.period not set before running task")
+            deadline += self._period
+            now = Timestamp.now()
+            while now >= deadline:
+                logger.warning("Missed deadline {}, it is now {}", deadline, now)
                 deadline += self._period
-                now = Timestamp.now()
-                while now >= deadline:
-                    logger.warn("Missed deadline {}, it is now {}", deadline, now)
-                    deadline += self._period
-
-                timeout = (deadline - now).s
-                await asyncio.wait_for(
-                    asyncio.shield(self._interval_task_stop_future), timeout=timeout
-                )
-                self._interval_task_stop_future.result()
-                logger.info("stopping IntervalSource task")
-                break
-            except asyncio.TimeoutError:
-                # This is the normal case, just continue with the loop
-                continue
-
-    async def stop(self, exception: Optional[Exception] = None) -> None:
-        logger.debug("stop()")
-        if self._interval_task_stop_future is not None:
-            self._interval_task_stop_future.set_result(None)
-        await super().stop(exception)
+
+            timeout = (deadline - now).s
+            done, _ = await asyncio.wait(
+                (
+                    asyncio.sleep(timeout),
```

Comment:
According to the documentation of `wait`, it is deprecated to pass in "coroutine objects" and unlike `gather`, which returns an `awaitable`, `sleep` returns a `coroutine`. 

Whatever that gibberish means, I guess `sleep` should be wrapped in a `create_task`. 🤷

Reply:
Please check fc20df5 that should address those three IntervalSource issues. However, I haven't been able to test it yet. Unfortunately the existing IntervalSource test doesn't cover this... Maybe we can improve the test but it might just be a case of manual tests :/.